base/compiler/typeinfer.jl

```diff
@@ -1070,27 +1117,9 @@ function typeinf_type(interp::AbstractInterpreter, mi::MethodInstance)
 end
 
 # This is a bridge for the C code calling `jl_typeinf_func()`
-typeinf_ext_toplevel(mi::MethodInstance, world::UInt) = typeinf_ext_toplevel(NativeInterpreter(world), mi)
-function typeinf_ext_toplevel(interp::AbstractInterpreter, mi::MethodInstance)
-    if isa(mi.def, Method)
-        # method lambda - infer this specialization via the method cache
-        src = typeinf_ext(interp, mi)
-    else
-        src = mi.uninferred::CodeInfo
-        if !src.inferred
```

Comment:
Do we not need to be careful to avoid recursion here anymore? Asking since in rare cases, the interpreter could theoretically decide to first infer and then codegen a toplevel thunk, and the codegen could choose to attempt to infer it again. I don't think the current code usually does that, but that is more of a policy decision than intended to be safe to assume

Reply:
I don't think there's any problem with that scenario. What issue are you imagining?

Reply:
(In particular, I don't think this changed. I don't believe the inference update to `inferred` was ever reflected here - at least in the current code).

Reply:
It seems to pass tests anyways, so I guess it seems not worth worrying about whether it might go wrong later.